storage/repair.go

```diff
@@ -227,66 +488,52 @@ func newDatabaseRepairer(database database) (databaseRepairer, error) {
 	}
 
 	r := &dbRepairer{
-		database:            database,
-		ropts:               ropts,
-		rtopts:              opts.RetentionOptions(),
-		shardRepairer:       shardRepairer,
-		repairStates:        make(map[time.Time]repairState),
-		sleepFn:             time.Sleep,
-		nowFn:               nowFn,
-		logger:              opts.InstrumentOptions().Logger(),
-		repairInterval:      ropts.RepairInterval(),
-		repairTimeOffset:    ropts.RepairTimeOffset(),
-		repairTimeJitter:    jitter,
-		repairCheckInterval: ropts.RepairCheckInterval(),
-		repairMaxRetries:    ropts.RepairMaxRetries(),
+		database:         database,
+		ropts:            ropts,
+		rtopts:           opts.RetentionOptions(),
+		shardRepairer:    shardRepairer,
+		repairStates:     make(map[time.Time]repairState),
+		nowFn:            nowFn,
+		logger:           opts.InstrumentOptions().Logger(),
+		repairInterval:   ropts.RepairInterval(),
+		repairTimeJitter: jitter,
+		repairMaxRetries: ropts.RepairMaxRetries(),
 	}
 	r.repairFn = r.Repair
 
 	return r, nil
 }
 
-func (r *dbRepairer) run() {
-	var curIntervalStart time.Time
-
-	for {
-		r.Lock()
-		closed := r.closed
-		r.Unlock()
-
-		if closed {
-			break
-		}
-
-		r.sleepFn(r.repairCheckInterval)
-
-		now := r.nowFn()
-		intervalStart := now.Truncate(r.repairInterval)
-
-		// If we haven't reached the offset yet, skip
-		target := intervalStart.Add(r.repairTimeOffset + r.repairTimeJitter)
-		if now.Before(target) {
-			continue
-		}
-
-		// If we are in the same interval, we must have already repaired, skip
-		if intervalStart == curIntervalStart {
-			continue
-		}
+func (r *dbRepairer) ShouldRun(start time.Time) bool {
+	if r.IsRepairing() {
+		return false
+	}
+	maxDelta := r.repairInterval + r.repairTimeJitter
+	return start.Sub(r.lastRun) > maxDelta
+}
 
-		curIntervalStart = intervalStart
-		if err := r.repairFn(); err != nil {
+func (r *dbRepairer) Run(start time.Time, mode runType) {
+	run := func() {
+		if err := r.repairFn(start); err != nil {
 			r.logger.Errorf("error repairing database: %v", err)
 		}
+		// setting lastRun regardless of success/failure
+		// to constrain the number of repairs run
+		r.lastRun = start
+	}
+
+	if mode == runTypeAsync {
+		go run()
+	} else {
+		run()
```

Comment:
Minor nit: should we standardize our usage and use the if == async in other places too?

Also this might be shorter to use everywhere:
```go
switch mode {
case runTypeAsync: go run()
default: run()
}
```